ophyd/status.py

```diff
@@ -55,35 +60,85 @@ def _timeout_thread(self):
             self._timeout_thread = None
 
     @_locked
-    def _finished(self, success=True, **kwargs):
+    def _finished(self, *, success=True, timestamp=None, **kwargs):
         # args/kwargs are not really used, but are passed - because pyepics
         # gives in a bunch of kwargs that we don't care about
-
         self.success = success
         self.done = True
-
-        if self._cb is not None:
-            self._cb()
-            self._cb = None
+        if timestamp is None:
+            timestamp = time.time()
+        self.finish_ts = timestamp
+        for cb in list(self._callbacks):
+            cb()
+            self._callbacks.remove(cb)
 
     @property
-    def finished_cb(self):
+    def callbacks(self):
         """
-        Callback to be run when the status is marked as finished
+        Callbacks to be run when the status is marked as finished
 
         The call back has no arguments
         """
-        return self._cb
+        return self._callbacks
 
-    @finished_cb.setter
     @_locked
-    def finished_cb(self, cb):
-        if self._cb is not None:
-            raise RuntimeError("Can not change the call back")
+    def add_callback(self, cb):
         if self.done:
             cb()
         else:
-            self._cb = cb
+            self._callbacks.add(cb)
+
+    def __and__(self, other):
+        """
+        Returns a new 'composite' status object, OrStatus,
+        with the same base API.
+
+        It will finish when both `self` or `other` finish.
+        """
+        return AndStatus(self, other)
+
+    @property
+    def elapsed(self):
+        if self.finish_ts is None:
+            return time.time() - self.start_ts
+        else:
+            return self.finish_ts - self.start_ts
+
+    def __repr__(self):
+        return '{0}(done={1.done}, elapsed={1.elapsed:.1f}, ' \
+               'success={1.success})'.format(self.__class__.__name__, self)
+
+
+class AndStatus(StatusBase):
+    "a Status that has composes two other Status objects using logical and"
+    def __init__(self, left, right, *, timestamp=None, timeout=None):
+        super().__init__(timestamp=timestamp, timeout=timeout)
+        self.left = left
+        self.right = right
+
+        def inner():
```

Comment:
Though the locks are `RLocks`, you're right, deadlock is a possibility here. A deadlock condition would depend on where left/right `finished` callbacks comes from. 

Let's see, if these two things happen around the same time:
1. Finished callback 1 comes from libca -> filtered through ophyd `MonitorDispatcher` to its Python thread -> status1._finished()
2. Finished callback 2 comes from some other Python thread -> status2._finished()

`status1` and `status2` run their callbacks simultaneously which correspond to `inner()` here. Assuming we attempt to lock both, that would be a deadlock.

OK, so it really depends on the callback implementation...